Add INotifyMoving interface

reaperrr opened the pull request:
And refactor some places to use it.

Dependency for #15746.

Fixes #16332.

reaperrr committed:
Add plumbing for notifying traits of movement

More precisely, about start and stop of movement.

pchote commented on OpenRA.Mods.Cnc/Activities/Leap.cs:
```diff
@@ -43,7 +45,9 @@ public Leap(Actor self, Actor target, Armament a, WDist speed, WAngle angle, Bit
 			mobile = self.Trait<Mobile>();
+			notifyMoving = self.TraitsImplementing<INotifyMoving>().ToArray();
 			mobile.SetLocation(mobile.FromCell, mobile.FromSubCell, targetMobile.FromCell, targetMobile.FromSubCell);
+			wasMoving = mobile.IsMoving;
```
Changing the actor state from inside the ctor is really bad practice (consider queuing).  Can we move these lines into `OnFirstRun` instead?

reaperrr replied:
Done. RA dogs still work, too.

pchote commented on OpenRA.Mods.Cnc/Activities/Leap.cs:
```diff
@@ -59,15 +63,30 @@ public Leap(Actor self, Actor target, Armament a, WDist speed, WAngle angle, Bit
 		{
-			if (ticks == 0 && IsCanceled)
-				return NextActivity;
+			if (ticks == 0)
```
Prefer `OnFirstRun` to explicitly counting ticks.

pchote reviewed (commented):
Code changes look generally reasonable, but I'd like to try my luck with requesting a scope creep:

pchote commented:
Tested that TD orca and TS tib fiends continue to work without regressions, so nominally :+1: depending on how the discussion above goes.

reaperrr commented:
Updated, and confimed that RA dog leaping still works.

pchote commented on OpenRA.Mods.Cnc/Activities/Leap.cs:
```diff
@@ -42,8 +44,8 @@ public Leap(Actor self, Actor target, Armament a, WDist speed, WAngle angle, Bit
 			this.damageTypes = damageTypes;
 			mobile = self.Trait<Mobile>();
+			notifyMoving = self.TraitsImplementing<INotifyMoving>().ToArray();
 			mobile.SetLocation(mobile.FromCell, mobile.FromSubCell, targetMobile.FromCell, targetMobile.FromSubCell);
```
Can you please try moving this stuff too?

reaperrr replied:
Just this line or the stuff below it as well?

pchote replied:
The stuff below too, as that depends on or changes the state of the actor at the time the activity first runs.

reaperrr replied:
Updated, but I kept the changes in a separate commit as the number of changes this required made me a bit uncomfortable about squashing it right away.

reaperrr commented:
Rebased.

abcdefg30 commented:
Needs another rebase. :/

reaperrr commented:
Rebased.

reaperrr commented:
This is a dependency for my upcoming animation priority system (because that requires getting rid of ReplaceAnim in WithMoveAnimation), and kind of for adding the long-requested turret and barrel equivalents of WithMoveAnimation (as it would be a waste of work to base them on bleed's approach).

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -46,6 +47,7 @@ public class Move : Activity
 		public Move(Actor self, CPos destination)
 		{
 			mobile = self.Trait<Mobile>();
+			notifyMoving = self.TraitsImplementing<INotifyMoving>().ToArray();
```
Tracking this in N places isn't ideal IMO.

Is there any reason why we can't do this inside `Mobile` by overloading the `IsMoving` setter?

```csharp
bool isMoving;
public bool IsMoving
{
	get
	{
		return isMoving;
	}
	set
	{
		if (value && !isMoving)
			foreach (var n in notifyMoving)
				n.StartedMoving(self);
		else if (!value && isMoving)
			foreach (var n in notifyMoving)
				n.StoppedMoving(self);

		isMoving = value;
	}
}
```

This would simplify the diff a lot, and make it a lot harder to accidentally forget to notify the interface consumers.

reaperrr replied:
This works for Mobile, but not so well for Aircraft.
Problem is that for aircraft, traits like WithMoveAnimation sometimes need to evaluate both IsMoving and IsMovingVertically from the notification methods. However, with the above approach, the IsMoving notifications are triggered before IsMovingVertically's value is updated, even if their value changes on the same tick. This can lead to wrong results.

So at least for `Aircraft`, I see no other way than taking the 'old' approach for now (but maybe with a dedicated method which also contains an explaining comment).

pchote replied:
Yeah, I only meant this for `Mobile`. Sorry if that wasn't clear.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnMovement.cs:
```diff
@@ -45,17 +45,59 @@ protected override void Created(Actor self)
 		}
 
-		void ITick.Tick(Actor self)
+		void INotifyMoving.StartedMoving(Actor self)
```
Having three different places that grant/revoke the condition doesn't sit too well with me.  Can we change this to have a single inner grant/revoke method that the other things call?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithMoveAnimation.cs:
```diff
@@ -46,17 +49,40 @@ public WithMoveAnimation(ActorInitializer init, WithMoveAnimationInfo info)
-			if (IsTraitDisabled || wsb.IsTraitDisabled)
-				return;
+			if (!IsTraitDisabled && !wsb.IsTraitDisabled)
+				wsb.PlayCustomAnimationRepeating(self, Info.MoveSequence);
```
Likewise here - it would be more robust if we could have a single inner method that plays/cancels, and is then called with whatever args are needed to distinguish the cases from the interface/enabled

pchote reviewed (commented):
A couple of ideas to try and reduce the chance of future regressions:

reaperrr commented:
Updated.

abcdefg30 commented:
This fails to compile, now.

reaperrr commented:
Updated again.
- Fixed compile issues (pushed too early)
- Made the Update methods in `WithMoveAnimation` and `GrantConditionOnMovement` more readable
- Updated `Aircraft` as mentioned in https://github.com/OpenRA/OpenRA/pull/15675#discussion_r263807008

pchote commented on OpenRA.Mods.Cnc/Activities/Leap.cs:
```diff
@@ -113,7 +115,9 @@ public override Activity Tick(Actor self)
-			mobile.IsMoving = true;
+			wasMoving = mobile.IsMoving;
+			if (!wasMoving)
+				mobile.IsMoving = true;
```
This shouldn't be needed - the `IsMoving` setter already checks the previous value and only notifies consumers if it changed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -321,12 +323,39 @@ protected virtual void Tick(Actor self)
 			cachedPosition = self.CenterPosition;
-			isMoving = (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0;
-			isMovingVertically = (oldCachedPosition - cachedPosition).VerticalLengthSquared != 0;
+			UpdateIsMoving(self, (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0, (oldCachedPosition - cachedPosition).VerticalLengthSquared != 0);
```
Style nit: this line is a bit too long IMO, so would be better to pull out `var horizontal = ...; var vertical = ...; UpdateIsMoving(self, horizontal, vertical);`.

pchote replied:
This probably no longer applies with my comment below.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -321,12 +323,39 @@ protected virtual void Tick(Actor self)
+			// were triggered first and evaluated IsMovingVertically before the latter has been updated, it would potentially make
+			// the wrong decision if both IsMoving and IsMovingVertically changed their value on the same tick.
+			// Therefore, updating both IsMoving *and* IsMovingVertically *before* any of the trait notifications are triggered is a must-have,
+			// at least for movement traits like Aircraft were both booleans play a role.
```
Considering this, IMO it really doesn't make sense to have separate notifications for horizontal vs vertical movement. Can we instead define a `[Flags] public enum MovementType { None, Horizontal, Vertical }` and then have one set of start/stop callbacks for both?

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnMovement.cs:
```diff
@@ -45,17 +45,66 @@ protected override void Created(Actor self)
-		void ITick.Tick(Actor self)
+		// playAttempt == called from TraitEnabled or StartedMoving(Vertically), else called from Stopped* or TraitDisabled
+		// vertical == called from one of the two *Vertically methods (or TraitEnabled), else called from some other method
+		void UpdateCondition(Actor self, bool grantAttempt = true, bool vertical = false)
```
The `MovementType` could then be passed in here instead of `vertical`

pchote replied:
There doesn't seem to be any code (except for the conditionManager check) shared between the `grantAttempt` true and false cases - best to have two methods for grant vs revoke.

pchote reviewed (commented):
Sorry to keep on scope creeping this, but i'm sure you agree its better to make sure this is as robust as possible before proceeding with things that depend on it.

reaperrr committed:
Adapt GrantConditionOnMovement to INotifyMoving

reaperrr committed:
Adapt WithMoveAnimation to INotifyMoving

And add ValidMovementTypes for configurability.

reaperrr commented:
Updated.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnMovement.cs:L59: [SA1119] The line contains unnecessary parenthesis.

reaperrr commented:
Fixed.

reaperrr committed:
MovementTypes refactor update rule

reaperrr commented:
Updated, polished the code and wrote update rule.

obrakmann reviewed (commented):
Looks ok to me. Didn't spot any in-game regressions. :+1:

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnMovement.cs:
```diff
@@ -45,17 +51,35 @@ protected override void Created(Actor self)
-			if (isMoving && conditionToken == ConditionManager.InvalidConditionToken)
-				conditionToken = conditionManager.GrantCondition(self, Info.Condition);
-			else if (!isMoving && conditionToken != ConditionManager.InvalidConditionToken)
+			var validMovement = !IsTraitDisabled
```
The goal here is to say "are any of the flags in `Info.ValidMovementTypes` set in `types`", right?

This can be written as the much simpler 
```csharp
var validMovement = !IsTraitDisabled && (types & Info.ValidMovementTypes) != 0;
```

The `&` operator returns 1 for each bit that is set in both `types` AND `Info.ValidMovementTypes`. So this will only return 0 overall if all the `MovementTypes` set in `types` are not set in `Info.ValidMovementTypes`, and visa versa.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -201,11 +215,31 @@ protected override void Created(Actor self)
+
+			// Check if actor has changed facing (turned). If yes, enable flag, if no, disable flag.
+			if (oldFacing != Facing)
+				CurrentMovementTypes |= MovementType.Turn;
```
As above, setting each flag in turn may cause the notification to be fired multiple times with different values in the same tick.  It would be best to set both flags in a temporary variable first, then assign that to `CurrentMovementTypes` once.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithMoveAnimation.cs:
```diff
@@ -34,29 +37,56 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
-			if (IsTraitDisabled || wsb.IsTraitDisabled)
+			var playAnim = false;
+			if (!IsTraitDisabled
+				&& ((considerHorizontal && types.HasFlag(MovementType.Horizontal))
```
This can simplify the same way as suggested above.

pchote reviewed (commented):
A few notes on how to simplify the bitwise logic:

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/20190314/ReplaceSpecialMoveConsiderations.cs:
```diff
@@ -0,0 +1,69 @@
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var mobileNode = actorNode.ChildrenMatching("Mobile").FirstOrDefault(m => !m.IsRemoval());
+			var considerTurnAsMove = false;
```
Autodetection will fail when `Mobile` is inherited, so this won't be reliable.  Can we change this to report uses of `GrantConditionOnMovement` and `WithMoveAnimation` for the modder to check and manually update instead?

pchote reviewed (commented):
Just two final comments, otherwise LGTM

reaperrr commented:
Updated.
I've tested the changes to the update rule too, works as intended.

reaperrr commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -188,18 +188,10 @@ public override Activity Tick(Actor self)
-				// or invisible mini-turns (due to less sprite facings than internal facings), we set IsMoving to  'true' during Turn activity
-				// when the facing delta is low enough to be covered with a single Turn tick.
-				// To avoid issues, we also make these mini-turns uninterruptible (like MovePart activities) to ensure the actor
+				// To avoid issues, we make 1-tick mini-turns uninterruptible (like MovePart activities) to ensure the actor
```
Done.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/cb9e4678dc76c587ef9f9e7bba82fb8018f5030b
